File: impls/java/src/mal/step1_read_print.java
package mal;

import mal.types.MalType;

public class step1_read_print {

	public MalType eval(MalType input) {
		return input;
	}

	public String rep(String input) {
		try {
			MalType afterRead = reader.read_str(input);
			MalType afterEval = this.eval(afterRead);
			return printer.pr_str(afterEval);
		} catch (Exception e) {
			return "EOF"; // generic error
		}
	}

	public static void main(String... args) {
		step1_read_print rp = new step1_read_print();

		// Local Test
		// System.out.println(rp.rep("\"abc\""));

		String input;
		do {
			input = System.console().readLine("user> ");
			if (input != null) {
				System.out.println(rp.rep(input));
			}
		} while (input != null);

	}
}

File: impls/clojure/src/mal/step1_read_print.clj
(ns mal.step1_read_print
  (:require [mal.reader :as reader]
            [mal.printer :as printer]))

(def READ reader/read_str)
(def EVAL identity)
(def PRINT printer/pr_str)

(def rep (comp PRINT EVAL READ))

(defn prompt []
  (print "user> ")
  (flush)
  (read-line))

;; MAIN LOOP

(defn -main [& args]
  (loop [input (prompt)]
    (when input
      (try
        (println (rep input))
        (catch Exception ex
          (when-let [message (ex-message ex)] ;; print message only when there's one
            (println message))))
      (recur (prompt)))))

